### PR TITLE
feat: add referer-restriction plugin

### PR DESCRIPTION
**Pull request by spacewander**
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: t/plugin/referer-restriction.t [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                    whitelist = {c}
+                })
+                if ok then
+                    ngx.log(ngx.ERR, c)
+                end
+            end
+        }
+    }
+--- request
+GET /t
```
Comment (nic-chen) [inline t/plugin/referer-restriction.t:189]: has no assert here？

Comment (spacewander) [reply t/plugin/referer-restriction.t:189]: @nic-chen 
All tests under this file already check the error log.

Comment (nic-chen) [reply t/plugin/referer-restriction.t:189]: OK

File: apisix/plugins/referer-restriction.lua [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+local http      = require "resty.http"
+local lrucache  = core.lrucache.new({
+    ttl = 300, count = 512
+})
+
+
+local schema = {
+    type = "object",
+    properties = {
+        optional = {
```
[the remark this answers is not in the document]
Comment (nic-chen) [reply apisix/plugins/referer-restriction.lua]: It’s still a bit confused, maybe we could use ʻallow_no_referer` or something. just a suggestion

**Review by nic-chen — APPROVED** [review]

**Comment by membphis** [comment]
need to update those docs too(add reference):

https://github.com/apache/apisix/blob/master/README.md
https://github.com/apache/apisix/blob/master/doc/README.md
https://github.com/apache/apisix/blob/master/doc/_sidebar.md

**Comment by spacewander** [comment]
@membphis 
Done

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@spacewander merged, many thx

**Pull request merged**